[Python] fixed subscription crash

tianfeng-yang opened the pull request:
Fixed #32310
1. The end of the `ReadClient` life cycle is `OnDone`. If `OnError` happend, `OnDone` will eventually be executed. In the existing Python code, `delete it` is executed in `OnDone`. At this time, `pCallback` and `pReadClient` (moved from `pychip_ReadClient_Read`) have both been deleted. , After that, if the python application executes `SubscriptionTransaction.Shutdown()` or `delete`, `pychip_ReadClient_Abort` will be executed, which will cause `pCallback` to be deleted again, causing the application to crash. Therefore, when calling `Shuddown()` or `delete` on the python side, you need to determine whether `OnError` happend. If `OnError` happend, you do not need to call `pychip_ReadClient_Abort` again. If not, it is a normal delete behavior on the application side.
2. The destructor of `ReadClient` also describes that after deleting `ReadClient`, `OnDone` will not be executed, so we can directly delete `ReadClient` in the callback of `OnDone`(the life cycle of `ReadClient` has ended), so that there is no need to move the memory of `pReadClient` to `ReadClientCallback` in `pychip_ReadClient_Read`.    
@Notes: There is no problem moving `ReadClient` to `pCallback` here, but `WriteClientCallback` is also not moved, just for consistency reasons.

[Follow the advice](https://github.com/project-chip/connectedhomeip/pull/32257#issuecomment-2061738561)

The following two points are split into PR #32308

1. Subscription support host test mode, align to `chip-tool`
2. Export `GetSubscriptionTimeoutMs` interface to check if the subscription is alive.

andy31415 commented on src/controller/python/chip/clusters/Attribute.py:
```diff
@@ -520,6 +520,22 @@ def GetReportingIntervalsSeconds(self) -> Tuple[int, int]:
 
         return minIntervalSec.value, maxIntervalSec.value
 
+    def GetSubscriptionTimeout(self) -> int:
```
Please add unit of measurement. `GetSubscriptionTimeoutMs` and describe it in the documentation as well.

Otherwise typical timeouts in python for things like sleep is "float as seconds"

tianfeng-yang replied:
Ok

andy31415 commented on src/controller/python/chip/clusters/attribute.cpp:
```diff
@@ -225,23 +225,25 @@ class ReadClientCallback : public ReadClient::Callback
+        // SubscriptionType will delete with pychip_ReadClient_Abort
+        if (apReadClient->IsReadType())
+        {
+            delete apReadClient;
```
the change of not having apReadClient as a member of the class anymore is not described in the summary of the PR.

Please explain changes and why in the summary of the PR.

tianfeng-yang replied:
Has been restored here

yunhanw-google commented on src/controller/python/chip/clusters/attribute.cpp:
```diff
@@ -225,23 +225,25 @@ class ReadClientCallback : public ReadClient::Callback
 
-        delete this;
+        // SubscriptionType will delete with pychip_ReadClient_Abort
+        if (apReadClient->IsReadType())
```
onDone should take care of the readclient release. Abort is called from python application, and it is not called from IM's readClient, when subscription onError is called after backoff expired, onDone is further called, then memory leak for readClient would happen.

tianfeng-yang replied:
In the original design, when a subscription `OnError` occurs, the memory is released in `OnDone`. If `pychip_ReadClient_Abort` is called again at this time, it will cause a secondary release and cause a crash.

yunhanw-google commented on src/controller/python/chip/clusters/attribute.cpp:
```diff
@@ -472,6 +475,22 @@ PyChipError pychip_ReadClient_GetReportingIntervals(ReadClient * pReadClient, ui
     return ToPyChipError(err);
 }
 
+void pychip_ReadClient_GetSubscriptionTimeout(ReadClient * pReadClient, uint32_t * milliSec)
```
when you find timeout happens, the existing abort should be enough for the callback/readclient destroy

tianfeng-yang replied:
On the contrary, calling `pychip_ReadClient_Abort` after OnError occurs will result in repeated release of memory

yunhanw-google replied:
I think you should not call pychip_ReadClient_Abort when onError happens....

yunhanw-google replied:
if onError happens, you call abort from application to release callback, then callback's onDone would be problematic...

tianfeng-yang replied:
According to your idea, the python interface needs to be modified. Now `pychip_ReadClient_Abort` is called in `Shutdown`, which will cause a crash when it is released normally after use on the application side.

tianfeng-yang replied:
> if onError happens, you call abort from application to release callback, then callback's onDone would be problematic...in readclient, ondone is called immediately after onerror..

Therefore, add ReadType judgment in `OnDone`. If it is SubscriptionType, hand it over to Abort to release it.

tcarmelveilleux commented on src/controller/python/chip/clusters/Attribute.py:
```diff
@@ -585,10 +601,13 @@ def Shutdown(self):
-                self._readTransaction._pReadClient, self._readTransaction._pReadCallback))
+        # Abort only when no error occurs.
+        # If an error occurs, pReadClient has been deleted internally.
+        if self._readTransaction._resultError is None:
```
Please open an issue showing what is currently failing, so that it's clear what is being fixed.

tianfeng-yang replied:
refer to the first point of comment

tcarmelveilleux reviewed (changes requested):
- [ ] Please open issue related to what the problem is with pychip_ReadClient_Abort
- [ ] Better explain what is the current problem  with callbacks, given that large change to resource management are being done. A failing test case being documented in the issue, and added in this PR in a way that it now passes, with no change to the test, but just changes to the code, would be helpful.
- [ ] Consider splitting the other non-contentious bits to a different PR (e.g. access to GetSubscriptionTimeout).
- [ ] Indicate how this was tested to avoid regressions in certification test scripts

tianfeng-yang commented:
> * [ ]  Please open issue related to what the problem is with pychip_ReadClient_Abort
> * [ ]  Better explain what is the current problem  with callbacks, given that large change to resource management are being done. A failing test case being documented in the issue, and added in this PR in a way that it now passes, with no change to the test, but just changes to the code, would be helpful.
> * [ ]  Consider splitting the other non-contentious bits to a different PR (e.g. access to GetSubscriptionTimeout).
> * [ ]  Indicate how this was tested to avoid regressions in certification test scripts

Ok, I've split the PR as you suggested and created an Issue for the controversial parts

tianfeng-yang commented:
@tcarmelveilleux I have modified it according to your suggestion. Are there any problems now?

tcarmelveilleux commented on src/controller/python/chip/clusters/attribute.cpp:
```diff
@@ -455,6 +453,7 @@ void pychip_ReadClient_Abort(ReadClient * apReadClient, ReadClientCallback * apC
     VerifyOrDie(apReadClient != nullptr);
     VerifyOrDie(apCallback != nullptr);
 
+    delete apReadClient;
```
This can lead to double-free if Abort is ever called after OnDone. Was this validated? This behavior existed before, however.

tianfeng-yang replied:
It will not be double-free. The subscription will only execute `OnDone` when an error occurs. When calling `pychip_ReadClient_Abort`, it has been confirmed whether an error occurred.

yunhanw-google commented:
@tianfeng-yang I would suggest to use ShutdownSubscriptions from https://github.com/project-chip/connectedhomeip/blob/master/src/app/InteractionModelEngine.cpp#L296, and remove abort API. and leave the below unchanged.
    *pReadClient = readClient.get();
    *pCallback   = callback.get();

    callback->AdoptReadClient(std::move(readClient));

    callback.release();

tianfeng-yang commented:
> @tianfeng-yang I would suggest to use ShutdownSubscriptions from https://github.com/project-chip/connectedhomeip/blob/master/src/app/InteractionModelEngine.cpp#L296, and remove abort API. and leave the below unchanged.
> 
> ```
>     *pReadClient = readClient.get();
>     *pCallback   = callback.get();
> 
>     callback->AdoptReadClient(std::move(readClient));
> 
>     callback.release();
> ```

Changing the API like this is a bit extensive, and I can't confirm whether it will cause a chain reaction.

tianfeng-yang committed:
[Python] After SubscriptionTransaction has an error, calling Shutdown() will crash

tianfeng-yang commented:
[Follow the advice](https://github.com/project-chip/connectedhomeip/pull/32257#issuecomment-2061738561)

tianfeng-yang committed:
Add a comment

tianfeng-yang committed:
update comment

andy31415 commented:
NOTE on old PR staleness review: this PR seems to be under active development and could use some reviews.
